azurelinuxagent/common/osutil/default.py

```diff
@@ -669,49 +672,71 @@ def get_if_mac(self, ifname):
+        Return the sizeof struct ifinfo. On 64-bit platforms the size is 40 bytes;
+        on 32-bit platforms the size is 32 bytes.
         """
-        iface=''
-        expected=16 # how many devices should I expect...
-
-        # for 64bit the size is 40 bytes
-        # for 32bit the size is 32 bytes
         python_arc = platform.architecture()[0]
         struct_size = 32 if python_arc == '32bit' else 40
+        return struct_size
 
-        sock = socket.socket(socket.AF_INET,
-                             socket.SOCK_DGRAM,
-                             socket.IPPROTO_UDP)
-        buff=array.array('B', b'\0' * (expected * struct_size))
-        param = struct.pack('iL',
-                            expected*struct_size,
-                            buff.buffer_info()[0])
-        ret = fcntl.ioctl(sock.fileno(), 0x8912, param)
-        retsize=(struct.unpack('iL', ret)[0])
-        if retsize == (expected * struct_size):
+
+    def _get_all_interfaces(self):
+        """
+        Return a dictionary mapping from interface name to IPv4 address.
+        Interfaces without a name are ignored.
+        """
+        expected=16 # how many devices should I expect...
+        struct_size = DefaultOSUtil._get_struct_ifconf_size()
+        array_size = expected * struct_size
+
+        buff = array.array('B', b'\0' * array_size)
+        param = struct.pack('iL', array_size, buff.buffer_info()[0])
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+        ret = fcntl.ioctl(sock.fileno(), IOCTL_SIOCGIFCONF, param)
+        retsize = (struct.unpack('iL', ret)[0])
+        if retsize == array_size:
             logger.warn(('SIOCGIFCONF returned more than {0} up '
                          'network interfaces.'), expected)
-        sock = buff.tostring()
-        primary = bytearray(self.get_primary_interface(), encoding='utf-8')
-        for i in range(0, struct_size * expected, struct_size):
-            iface=sock[i:i+16].split(b'\0', 1)[0]
-            if len(iface) == 0 or self.is_loopback(iface) or iface != primary:
-                # test the next one
-                if len(iface) != 0 and not self.disable_route_warning:
-                    logger.info('Interface [{0}] skipped'.format(iface))
-                continue
-            else:
-                # use this one
-                logger.info('Interface [{0}] selected'.format(iface))
-                break
+        sock = None
+        ifconf_buff = buff.tostring()
+
+        ifaces = {}
+        for i in range(0, array_size, struct_size):
+            iface = ifconf_buff[i:i+16].split(b'\0', 1)[0]
+            if len(iface) > 0:
+                iface_name = iface.decode('latin-1')
+                if iface_name not in ifaces:
+                    ifaces[iface_name] = socket.inet_ntoa(ifconf_buff[i+20:i+24])
+        return ifaces
+
+
+    def get_first_if(self):
+        """
+        Return the interface name, and ip addr of the
```

Comment:
NIT: ip -> IPv4

Reply:
I updated this entire docstring to reflect the actual goal (return the IPv4 of the primary interface; if not available, the IPv4 of any interface).